src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.SslCtx.cs

```diff
@@ -69,7 +83,107 @@ protected override bool ReleaseHandle()
         {
             Interop.Ssl.SslCtxDestroy(handle);
             SetHandle(IntPtr.Zero);
+            if (_gch.IsAllocated)
+            {
+                //Interop.Ssl.SslCtxSetData(this, (IntPtr)_gch);
+                _gch.Free();
+            }
+
+            if (_sslSessions != null)
+            {
+                lock (_sslSessions)
+                {
+                    foreach (string name in _sslSessions.Keys)
+                    {
+                        _sslSessions.Remove(name, out IntPtr session);
+                        Interop.Ssl.SessionFree(session);
+                    }
+                }
+            }
+
             return true;
         }
+
+        public void EnableSessionCache()
+        {
+            _sslSessions = new ConcurrentDictionary<string, IntPtr>();
+            _gch = GCHandle.Alloc(this);
+            // This is needed so we can find the handle from session remove callback.
+            Interop.Ssl.SslCtxSetData(this, (IntPtr)_gch);
+        }
+
+        public bool TryAddSession(IntPtr serverName, IntPtr session)
+        {
+            Debug.Assert(_sslSessions != null && session != IntPtr.Zero);
+
+            if (_sslSessions == null || serverName == IntPtr.Zero)
+            {
+                return false;
+            }
+
+            string? name = Marshal.PtrToStringAnsi(serverName);
+            if (!string.IsNullOrEmpty(name))
+            {
+                Interop.Ssl.SessionSetHostname(session, serverName);
```

Comment:
It feels really weird to me that SetHostname would be inside TryAddSession.

Reply:
I need something to find the session in removal. Associating name with it allows me to get the string and then do lookup. It would be great if we can come up with something that allows to lookup by both IntPtr and Name.

Reply:
Sure... it just feels like calling SetHostname'd be the responsibility of the caller.  Doesn't it have to be done in the case where TryAddSession returns false?

Reply:
This is ony only for the lookup - there is no functional difference. And I tried to hide all this inside the handle. 
If we can lookup/remove the entry just from the IntPtr session, we would not need to do that. But I'm not sure if there is good way.

Reply:
I put in comment and made changes to make it clear. I also moved complementing removal so both parts are done inside the SafeHandle.